doc/optimal_power_flow.rst

```diff
@@ -639,19 +641,118 @@ additional shadow prices for constraints.
+All variable and constraint references are stored in the network object itself, attached to the relevant component. By accessing these references inside an ``extra_functionality(network, snapshots)`` function passed to ``network.lopf``, you can select variables and build constraints using the following functions:
 
-* :py:meth:`pypsa.linopt.get_var` for getting the variables which should be included in the constraint
+
+* :py:meth:`pypsa.linopt.get_var` for getting the variables which should be included in the constraint.
 * :py:meth:`pypsa.linopt.linexpr` for creating linear expressions for the left hand side (lhs) of the constraint. Note that only the lhs includes all terms with variables, the rhs is a constant.
+* :py:meth:`pypsa.linopt.join_expr` for summing linear expressions.
 * :py:meth:`pypsa.linopt.define_constraints` for defining a network constraint.
 
+Once the problem has been built, all names of variable sets are stored in ``n.variables`` and all names of constraint sets in ``n.constraints``.
 
 The function ``extra_postprocessing`` is not necessary when pyomo is deactivated. For retrieving additional shadow prices, just pass the name of the constraint, to which the constraint is attached, to the ``keep_shadowprices`` parameter of the ``lopf`` function.
 
+
+get_var
+^^^^^^^
+
+The function ``linopt.get_var`` is used to access the variables attached to a component. To find out which variables are available, look inside ``n.variables`` once the ``n.lopf(pyomo=False)`` has run.
+
+For example, to access the the dispatch variable of ``network.generators_t.p`` attached to the component ``Generator`` use
+
+  >>> get_var(n, 'Generator', 'p')
+
+This will return a ``pd.DataFrame`` with index of ``network.snapshots`` and columns of ``network.generators.index`` with the variable references in each entry.
+
+To access the capacities of extendable generators use
+
+  >>> get_var(n, 'Generator', 'p_nom')
+
+This will return a ``pd.Series`` with index of ``network.generators.index`` with the variable references in each entry.
+
+linexpr
+^^^^^^^
+
+The function ``linopt.linexpr`` is used to build linear combinations of variables.
+
+It takes a tuple of twoples, where the first entry is the coefficients and the second entry is the variables.
+
+Beware that the indices and columns of the ``pd.DataFrame`` or  ``pd.Series`` you combine must have aligned indices and columns. This applies both to coefficients and variables.
+
+For example, to subtract the generator capacity from the dispatch do
+
+  >>> linexpr((1, get_var(n, 'Generator', 'p')), (-1, get_var(n, 'Generator', 'p_nom')))
+
+TODO: this doesn't work since the indices don't align.
+
+To add the dispatch weighted by the snapshot weightings do
+
+  >>> linexpr((n.snapshot_weightings['weightings'], get_var(n, 'Generator', 'p')), (-1, get_var(n, 'Generator', 'p_nom')))
+
+TODO: this doesn't work since the indices don't align.
```

Comment:
This does not work as series as broadcasted along the index axis when applying an operation on a dataframe (linopy handles such cases better :) )
```suggestion
  >>>  ext_i = n.get_extendable_i(n, 'Generator')
  >>>  p = get_var(n, 'Generator', 'p').loc[ext_i]
  >>>  weightings = pd.DataFrame({gen: n.snapshot_weightings.generators for gen in ext_i})
  >>>  p_nom = get_var(n, 'Generator', 'p_nom')
  >>>  linexpr((weightings, p), (-1, p_nom))
```